ip_country/src/country_block_stream.rs

```diff
@@ -0,0 +1,411 @@
+pub struct CountryBlock {
+    pub ip_range: IpRange,
+    pub country: Country,
+}
+
+impl TryFrom<StringRecord> for CountryBlock {
+    type Error = String;
+
+    fn try_from(string_record: StringRecord) -> Result<Self, Self::Error> {
+        let mut iter = string_record.iter();
+        let start_ip = Self::ip_addr_from_iter(&mut iter)?;
+        let end_ip = Self::ip_addr_from_iter(&mut iter)?;
+        let iso3166 = match iter.next() {
+            None => return Err("CSV line contains no ISO 3166 country code".to_string()),
+            Some(s) => s,
+        };
+        if iter.next().is_some() {
+            return Err(format!(
+                "CSV line should contain 3 elements, but contains {}",
+                string_record.len()
+            ));
+        };
+        Self::validate_ip_addresses(start_ip, end_ip)?;
+        let country = Country::try_from(iso3166)?;
+        let country_block = match (start_ip, end_ip) {
+            (IpAddr::V4(start), IpAddr::V4(end)) => CountryBlock {
+                ip_range: IpRange::V4(start, end),
+                country,
+            },
+            (IpAddr::V6(start), IpAddr::V6(end)) => CountryBlock {
+                ip_range: IpRange::V6(start, end),
+                country,
+            },
+            (start, end) => panic!(
+                "Start and end addresses must be of the same type, not {} and {}",
+                start, end
+            ),
+        };
+        Ok(country_block)
+    }
+}
+
+impl CountryBlock {
+    fn ip_addr_from_iter(iter: &mut StringRecordIter) -> Result<IpAddr, String> {
+        let ip_string = match iter.next() {
+            None => return Err("Missing IP address in CSV record".to_string()),
+            Some(s) => s,
+        };
+        let ip_addr = match IpAddr::from_str(ip_string) {
+            Err(e) => {
+                return Err(format!(
+                    "Invalid ({:?}) IP address in CSV record: '{}'",
+                    e, ip_string
+                ))
+            }
+            Ok(ip) => ip,
+        };
+        Ok(ip_addr)
+    }
+
+    fn validate_ip_addresses(start_ip: IpAddr, end_ip: IpAddr) -> Result<(), String> {
+        match (start_ip, end_ip) {
+            (IpAddr::V4(start_v4), IpAddr::V4(end_v4)) => {
+                if u32::from(start_v4) > u32::from(end_v4) {
+                    Err(format!(
+                        "Ending address {} is less than starting address {}",
+                        end_v4, start_v4
+                    ))
+                } else {
+                    Ok(())
+                }
+            }
+            (IpAddr::V6(start_v6), IpAddr::V6(end_v6)) => {
+                if u128::from(start_v6) > u128::from(end_v6) {
+                    Err(format!(
+                        "Ending address {} is less than starting address {}",
+                        end_v6, start_v6
+                    ))
+                } else {
+                    Ok(())
+                }
+            }
```

Comment:
Can be simplified by a reusable piece of function. 

```Rust
fn validate_ips_are_sequential<SingleIntegerIPRep, IP>(start: IP, end: IP)
-> Result<(), String> 
where 
  SingleIntegerIPRep: From<IP>,
  Ip: Display
{
      if SingleIntegerIPRep::from(start) > SingleIntegerIPRep::from(end) {
          Err(format!(
              "Ending address {} is less than starting address {}",
              end, start
          ))
      } else {
          Ok(())
      }
}
```
Once you have this function, use it twice, for both the IPv6 and IPv4.